src/lib.rs

```diff
@@ -16,168 +17,256 @@ use std::{
     funcs: Arc<bindings::metal_irconverter>,
 }
 
 impl Drop for IRObject {
+    #[doc(alias = "IRObjectDestroy")]
     fn drop(&mut self) {
-        unsafe { self.funcs.IRObjectDestroy(self.me) }
+        unsafe { self.funcs.IRObjectDestroy(self.me.as_ptr()) }
     }
 }
 
 impl IRObject {
-    pub fn create_from_dxil(
-        compiler: &IRCompiler,
-        bytecode: &[u8],
-    ) -> Result<IRObject, Box<dyn std::error::Error>> {
+    #[doc(alias = "IRObjectCreateFromDXIL")]
+    pub fn create_from_dxil(compiler: &IRCompiler, bytecode: &[u8]) -> IRObject {
         unsafe {
-            let me = compiler.funcs.IRObjectCreateFromDXIL(
+            let me = NonNull::new(compiler.funcs.IRObjectCreateFromDXIL(
                 bytecode.as_ptr(),
                 bytecode.len(),
-                bindings::IRBytecodeOwnership::None,
-            );
+                bindings::IRBytecodeOwnership::Copy,
+            ))
+            .expect("Failed to create IRObject from DXIL");
 
-            Ok(Self {
+            Self {
                 me,
                 funcs: compiler.funcs.clone(),
-            })
+            }
+        }
+    }
+
+    #[doc(alias = "IRObjectGatherRaytracingIntrinsics")]
+    pub fn gather_raytracing_intrinsics(&self, entry_point: &CStr) -> u64 {
+        unsafe {
+            self.funcs
+                .IRObjectGatherRaytracingIntrinsics(self.me.as_ptr(), entry_point.as_ptr())
         }
     }
 
-    pub fn get_type(&self) -> ffi::IRObjectType {
-        unsafe { self.funcs.IRObjectGetType(self.me) }
+    #[doc(alias = "IRObjectGetType")]
+    pub fn object_type(&self) -> ffi::IRObjectType {
+        unsafe { self.funcs.IRObjectGetType(self.me.as_ptr()) }
     }
 
-    pub fn get_metal_ir_shader_stage(&self) -> ffi::IRShaderStage {
-        unsafe { self.funcs.IRObjectGetMetalIRShaderStage(self.me) }
+    #[doc(alias = "IRObjectGetMetalIRShaderStage")]
+    pub fn metal_ir_shader_stage(&self) -> ffi::IRShaderStage {
+        unsafe { self.funcs.IRObjectGetMetalIRShaderStage(self.me.as_ptr()) }
     }
 
-    pub fn get_metal_lib_binary(
+    #[doc(alias = "IRObjectGetMetalLibBinary")]
+    pub fn metal_lib_binary(
         &self,
+        compiler: &IRCompiler,
         shader_stage: ffi::IRShaderStage,
-        dest_lib: &mut IRMetalLibBinary,
-    ) -> bool {
-        unsafe {
-            self.funcs
-                .IRObjectGetMetalLibBinary(self.me, shader_stage, dest_lib.me)
+    ) -> Result<IRMetalLibBinary, MetalLibNoBytecodeFoundError> {
+        let mtl_lib = IRMetalLibBinary::new(compiler);
+        if unsafe {
+            self.funcs.IRObjectGetMetalLibBinary(
+                self.me.as_ptr(),
+                shader_stage,
+                mtl_lib.me.as_ptr(),
+            )
+        } {
+            Ok(mtl_lib)
+        } else {
+            let obj_type = unsafe { self.funcs.IRObjectGetType(self.me.as_ptr()) };
+            let stage = unsafe { self.funcs.IRObjectGetMetalIRShaderStage(self.me.as_ptr()) };
```

Comment:
What's different about the return from `IRObjectGetMetalIRShaderStage()` versus the `shader_stage` argument? Are we assuming that it's different?

Reply:
If so we shouldn't let the user pass it, and just use the result from this function inside `IRObjectGetMetalLibBinary()`.

Reply:
I feel like we could either remove the `shader_stage` argument, or use the argument to verify the shader stage is actually what the user expects it to be. What do you think?
Also, I just realized those unsafe blocks can be replaced by `self.object_type()` etc.